File: HUBSPOT_TO_S3/csharp/hubspot2s3/HubSpot.cs
using System;
using System.Collections.Generic;
using RestSharp;

namespace hubspot2s3
{
    public class HubSpot
    {

        private static readonly NLog.Logger s_logger = NLog.LogManager.GetCurrentClassLogger();
        private readonly string _headers = "";
        private readonly string _queryString = "";

        public HubSpot(string HubSpotApiKey, string HubSpotAccessToken)
        {
            if (HubSpotApiKey != "")
            {
                _queryString = HubSpotApiKey;
            }
            else if (HubSpotAccessToken != "")
            {
                _headers = $"Bearer {HubSpotAccessToken}";
            }
        }

        public static string RelativeEntityUrl(string Entity, string Key)
        {
            Dictionary<string, string> _v1 = new Dictionary<string, string>()
            {
                { "contacts", "contact" }
            };
            Dictionary<string, string> _v2 = new Dictionary<string, string>()
            {
                {  "companies", "companies" }
            };

            string _rtn = "";

            if (_v1.ContainsKey(Entity))
            {
                _rtn = $"{Entity}/v1/{_v1[Entity]}";
            }
            else
            {
                _rtn = $"{Entity}/v2/{_v2[Entity]}";
            }

            if (Key != "")
            {
                _rtn += $"/{Key}";
            }

            return _rtn;
        }

        public string AbsoluteUrl(string Relative)
        {
            return $"https://api.hubapi.com/{Relative}";
        }

        public string GetFqdn(string Entity, string Key)
        {
            return AbsoluteUrl(RelativeEntityUrl(Entity, Key));
        }

        private IRestResponse Get(string Url, string Data)
        {
            var _client = new RestClient();
            var _request = new RestRequest(Url, Method.GET);
            _request.AddQueryParameter("hapikey", _queryString);
            _request.AddHeader("Content-Type", "application/json");
            _request.AddHeader("Authorization", _headers);

            return _client.Execute(_request);
        }

        private IRestResponse Post(string Url, string Data)
        {
            var _client = new RestClient();
            var _request = new RestRequest(Url, Method.POST);
            _request.AddQueryParameter("hapikey", _queryString);
            _request.AddHeader("Content-Type", "application/json");
            _request.AddHeader("Authorization", _headers);
            _request.AddJsonBody(Data);

            return _client.Execute(_request);
        }

        private IRestResponse Put(string Url, string Data)
        {
            var _client = new RestClient();
            var _request = new RestRequest(Url, Method.PUT);
            _request.AddQueryParameter("hapikey", _queryString);
            _request.AddHeader("Content-Type", "application/json");
            _request.AddHeader("Authorization", _headers);
            _request.AddJsonBody(Data);

            return _client.Execute(_request);
        }

        public IRestResponse Create(string Entity, string Data)
        {
            s_logger.Info($"Attempting to create a {Entity}");
            return Post(GetFqdn(Entity, ""), Data);

        }

        public IRestResponse Replace(string Entity, string Key, string Data)
        {
            s_logger.Info($"Attempting to replace a {Entity}/{Key}");
            return Put(GetFqdn(Entity, Key), Data);
        }
    }
}

File: HUBSPOT_TO_S3/java/src/main/java/hubspot2s3/HubSpot.java
package hubspot2s3;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;
import kong.unirest.HttpResponse;
import kong.unirest.JsonNode;
import kong.unirest.Unirest;

class HubSpot {

	private static Logger logger = Logger.getLogger(Main.class.getName());
	private String headers = "";
	private String queryString = "";

	HubSpot(String hubSpotApiKey, String hubSpotAccessToken) {
		if (hubSpotApiKey != "") {
			queryString = hubSpotApiKey;
		} else if (hubSpotAccessToken != "") {
			headers = "Bearer " + hubSpotAccessToken;
		}
	}

	public static String relativeEntityUrl(String entity, String key) {
		Map<String, String> v1 = new HashMap<String, String>();
		v1.put("contacts", "contact");
		Map<String, String> v2 = new HashMap<String, String>();
		v2.put("companies", "companies");

		String rtn = "";

		if (v1.containsKey(entity)) {
			rtn = entity + "/v1/" + v1.get(entity);
		} else {
			rtn = entity + "/v2/" + v2.get(entity);
		}

		if (key != "") {
			rtn += "/" + key;
		}

		return rtn;
	}

	public String absoluteUrl(String relative) {
		return "https://api.hubapi.com/" + relative;
	}

	public String getFqdn(String entity, String key) {
		System.out.println(relativeEntityUrl(entity, key));
		return absoluteUrl(relativeEntityUrl(entity, key));
	}

	private HttpResponse<JsonNode> get(String url, String data) {
		HttpResponse<JsonNode> response = Unirest.get(url).header("Content-Type", "application/json")
				.queryString("hapikey", queryString).header("Authorization", headers).asJson();

		return response;
	}

	private HttpResponse<JsonNode> post(String url, String data) {
		HttpResponse<JsonNode> response = Unirest.post(url).header("Content-Type", "application/json")
				.queryString("hapikey", queryString).header("Authorization", headers).body(data).asJson();

		return response;
	}

	private HttpResponse<JsonNode> put(String url, String data) {
		HttpResponse<JsonNode> response = Unirest.put(url).header("Content-Type", "application/json")
				.queryString("hapikey", queryString).header("Authorization", headers).body(data).asJson();
		return response;
	}

	public HttpResponse<JsonNode> create(String entity, String data) {
		logger.info("Attempting to create a " + entity);
		return post(getFqdn(entity, ""), data);

	}

	public HttpResponse<JsonNode> replace(String entity, String key, String data) {
		logger.info("Attempting to replace a " + entity + "/" + key);
		return put(getFqdn(entity, key), data);
	}
}
